### PR TITLE
Refactor cURL error handling to preserve response data

### LINKED ISSUE #32416: Improve logging on Manager
## Description

The current logging in the *indexer-connector* module is inconsistent and often lacks clear context. This makes it difficult to quickly understand what went wrong or what the system is doing. We need to enhance the logging so that *every* log emitted by the module is clear, actionable, and consistent.

This has been noticed during the support provided on

- https://github.com/wazuh/wazuh/issues/32044

## Definition of Done  
- [x] All log statements in the module provide enough context about the error and determinate the root cause.  
- [x] Tests or sample runs confirm that logs appear and are clear in scenarios (success, error, edge cases)

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
Closes https://github.com/wazuh/wazuh/issues/32416

# Summary

This change refactors the cURL execution logic to improve error handling and diagnostics. Specifically, it allows the `CurlException` to carry the HTTP response code while preserving the response body for inspection in case of errors.

## Problem Statement

Previously, the cURL handler would **reset the easy handle (`curl_easy_reset`) immediately after an error**, which cleared any buffered response data. This made it impossible to access the server’s response body in failure cases (e.g., 4xx/5xx responses).

Additionally, the use of `OPT_FAILONERROR` caused `curl_easy_perform()` to fail automatically on HTTP errors (≥ 400), without giving us a chance to read the full response.

**Observed behavior:**

* On HTTP 4xx/5xx responses, `curl_easy_perform()` returned `CURLE_HTTP_RETURNED_ERROR`.
* The handle was reset before the exception was thrown.
* The application lost the response body, preventing detailed error reporting or diagnostics.

**Expected behavior:**

* On failure, response data should still be available to the caller.
* The exception should contain the HTTP response code.
* The cURL handle should only be reset after successful execution.

## Solution

The `execute()` method in the cURL handler has been refactored to preserve the response context and provide richer exceptions.

### 1. Defer `curl_easy_reset()`

The reset now occurs **only on success**, ensuring the response data remains accessible when an exception is thrown:

```cpp
// Only reset on success
curl_easy_reset(m_curlHandler.get());
```

### 2. Preserve Response and HTTP Code on Errors

When `CURLE_HTTP_RETURNED_ERROR` occurs, the HTTP response code is retrieved and passed into the `CurlException`, allowing downstream consumers to handle errors based on status code:

```cpp
if (resPerform == CURLE_HTTP_RETURNED_ERROR)
{
    if (resGetInfo != CURLE_OK)
    {
        throw std::runtime_error("cURLSingleHandler::execute() failed: Couldn't get HTTP response code");
    }
    // Throw exception WITHOUT resetting - response data is still available
    throw Curl::CurlException(curl_easy_strerror(resPerform), responseCode);
}
```

### 3. Remove `OPT_FAILONERROR`

The following option is now commented out:

```cpp
// this->setOptionLong(OPT_FAILONERROR, 1l);
```

This prevents automatic failure on HTTP errors, letting the new `execute()` logic handle them manually while still capturing the response body.

## Changes Made

* **`curlSingleHandler.cpp`**

  * Moved `curl_easy_reset()` to only execute on successful requests.
  * Updated exception handling to include response codes and preserve body data.
  * Improved runtime error messages for missing response info.

* **`curlWrapper.hpp`**

  * Updated `CurlException` to accept an optional HTTP response code argument.

* **Removed automatic fail-on-error option (`OPT_FAILONERROR`)** to allow full response capture on server errors.

### CODE REVIEW

File: src/curlWrapper.hpp [hunk 1]
```diff
@@ -216,7 +220,15 @@ class cURLWrapper final : public IRequestImplementator
...
 
-        m_curlHandler->execute();
+        try
+        {
+            m_curlHandler->execute();
+        }
+        catch (Curl::CurlException& ex)
+        {
+            // Note: m_returnValue contains the response body, even for errors
+            throw Curl::CurlException(ex.what(), ex.responseCode(), m_returnValue);
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/curlWrapper.hpp]: I need this

File: shared/curlException.hpp [hunk 2]
```diff
@@ -18,58 +18,74 @@
...
 
-        private:
-            std::runtime_error m_error;
-            const long m_responseCode;
-    };
-}
+private:
+    std::runtime_error m_error;
+    const long m_responseCode;
+    std::string m_responseBody;
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply shared/curlException.hpp]: Fixed

File: src/curlSingleHandler.hpp [hunk 3]
```diff
@@ -53,7 +53,8 @@ class cURLSingleHandler final : public ICURLHandler
         long responseCode;
         const auto resGetInfo {curl_easy_getinfo(m_curlHandler.get(), CURLINFO_RESPONSE_CODE, &responseCode)};
 
-        curl_easy_reset(m_curlHandler.get());
+        // DON'T reset here anymore
+        // curl_easy_reset(m_curlHandler.get());
```
Comment (MarcelKemp) [inline src/curlSingleHandler.hpp]: If an error occurs, we would need the handler to be reset, because otherwise it could cause a residual state.

File: src/curlSingleHandler.hpp [hunk 4]
```diff
@@ -63,10 +64,14 @@ class cURLSingleHandler final : public ICURLHandler
                 {
                     throw std::runtime_error("cURLSingleHandler::execute() failed: Couldn't get HTTP response code");
                 }
+                // Throw exception WITHOUT resetting - response data is still available
```
Comment (MarcelKemp) [inline src/curlSingleHandler.hpp]: Can we unify the errors? So that if it is `resGetInfo != CURLE_OK`, it returns 0, while in other cases, it will return the HTTP error >0.

File: src/curlWrapper.hpp [hunk 5]
```diff
@@ -117,7 +117,11 @@ class cURLWrapper final : public IRequestImplementator
 
         this->setOptionPtr(OPT_WRITEDATA, &m_response);
 
-        this->setOptionLong(OPT_FAILONERROR, 1l);
+        // this->setOptionLong(OPT_FAILONERROR, 1l);
+
+        // Note: The OPT_FAILONERROR option makes cURL fail on HTTP response codes >= 400.
+        // However, in such cases we want to capture the response body, so we don't set it.
+        // Instead, we handle HTTP errors in the execute() method.
```
Comment (MarcelKemp) [inline src/curlWrapper.hpp]:
```suggestion
        // Note: OPT_FAILONERROR is intentionally NOT set. This option would cause cURL 
        // to fail automatically on HTTP response codes >= 400, preventing us from 
        // capturing the response body. We want to allow callers to handle HTTP errors 
        // with full access to the server's response.
```

File: src/curlSingleHandler.hpp (hunk 3 above)
Comment (MarcelKemp) [inline src/curlSingleHandler.hpp]:
```suggestion
        // Note: curl_easy_reset() is deferred to after successful execution (line 73)
        // to preserve response data for error inspection and exception construction.
```

**Review by MarcelKemp — CHANGES_REQUESTED** [review]

File: src/curlSingleHandler.hpp (hunk 3 above)
Comment (GabrielEValenzuela) [reply src/curlSingleHandler.hpp]: We can reset it

```console
[==========] Running 11 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 11 tests from IndexerConnectorTest
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_WithErrors
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_WithErrors (7219 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_MultipleErrors
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_MultipleErrors (7171 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_NoErrorsField
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_NoErrorsField (6633 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_ErrorsFalse
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_ErrorsFalse (6698 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_MissingErrorReason
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_MissingErrorReason (6756 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_MissingErrorType
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_MissingErrorType (6760 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_InvalidJSON
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_InvalidJSON (6747 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_ShardLimitExceeded
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_ShardLimitExceeded (6725 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_IndexNotFound
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_IndexNotFound (6740 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_Forbidden
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_Forbidden (6756 ms)
[ RUN      ] IndexerConnectorTest.HandleIndexerInternalErrors_ClusterUnavailable
[       OK ] IndexerConnectorTest.HandleIndexerInternalErrors_ClusterUnavailable (6750 ms)
[----------] 11 tests from IndexerConnectorTest (74963 ms total)

[----------] Global test environment tear-down
[==========] 11 tests from 1 test suite ran. (74963 ms total)
[  PASSED  ] 11 tests.
```

File: src/curlSingleHandler.hpp (hunk 4 above)
Comment (GabrielEValenzuela) [reply src/curlSingleHandler.hpp]: We can’t. The code test is tightly tied to the exceptions.

File: src/curlSingleHandler.hpp (hunk 3 above)
Comment (GabrielEValenzuela) [reply src/curlSingleHandler.hpp]: Done!

File: src/curlWrapper.hpp (hunk 5 above)
Comment (GabrielEValenzuela) [reply src/curlWrapper.hpp]: Done!

**Review by MarcelKemp — APPROVED** [review]
LGTM.
